### PR TITLE
Feature/add novuln bool to vulnerability filter

### PR DESCRIPTION
**Pull request by pxp928**
# Description of the PR

Followup to PR https://github.com/guacsec/guac/pull/1147

Adding the functionality back that was introduced in https://github.com/guacsec/guac/pull/1073 to allow for filtering of vulnerabilities (no vulnerability)

```
# Setting noVuln to true return the noVuln node
# Setting noVuln to false return all vulnerabilities nodes
query noVuln {
  vulnerabilities(vulnSpec: {noVuln: true}) {
    ...allVulnerabilityTree
  }
}
```

```
# Setting noVuln to true return all packages that have no vulnerabilities
# Setting noVuln to false return all packages with vulnerabilities
query CertifyNoVuln {
  CertifyVuln(certifyVulnSpec: {
    vulnerability: {noVuln: true}
  }) {
    ...allCertifyVulnTree
  }
}
```

# PR Checklist

- [x] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [x] All new changes are covered by tests
- [x] If GraphQL schema is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [x] All CI checks are passing (tests and formatting)
- [x] All dependent PRs have already been merged

### CODE REVIEW

File: pkg/assembler/graphql/schema/vulnerability.graphql [hunk 1]
```diff
@@ -70,11 +70,18 @@ Use null to match on all values at that level.
 For example, to get all vulnerabilities in GUAC backend, use a VulnSpec
 where every field is null.
 
+Setting the noVuln boolean true will ignore the other inputs for type and vulnerabilityID.
+Setting noVuln to true means retrieving only nodes where the type of the vulnerability is "novuln" 
+and the it has an empty string for vulnerabilityID. Setting it to false means retrieving only nodes
+with identified vulnerabilities. Setting one of the other fields and omitting the noVuln means retrieving
+vulnerabilities for the corresponding type and vulnerabilityID.
```
Comment (lumjjb) [inline pkg/assembler/graphql/schema/vulnerability.graphql:77]: from what i read here false == unset, is that true?

Comment (lumjjb) [reply pkg/assembler/graphql/schema/vulnerability.graphql:77]: can i suggest
`Setting it to false filters out all results that are "novuln".`

Comment (pxp928) [reply pkg/assembler/graphql/schema/vulnerability.graphql]: updated.

File: pkg/assembler/backends/inmem/certifyVuln.go [hunk 2]
```diff
@@ -212,6 +213,21 @@ func (c *demoClient) CertifyVuln(ctx context.Context, filter *model.CertifyVulnS
...
+			VulnerabilityID: ptrfrom.String(""),
+		})
+		if err != nil {
+			return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+		}
+		if exactVuln != nil {
+			search = append(search, exactVuln.certifyVulnLinks...)
+			foundOne = true
+		}
+	}
```
Comment (lumjjb) [inline pkg/assembler/backends/inmem/certifyVuln.go:230]: this should be else if - we don't need to go into the next branch if it is novuln

File: pkg/assembler/backends/inmem/certifyVuln.go [hunk 3]
```diff
@@ -212,6 +213,21 @@ func (c *demoClient) CertifyVuln(ctx context.Context, filter *model.CertifyVulnS
...
 	if !foundOne && filter != nil && filter.Vulnerability != nil {
 		exactVuln, err := c.exactVulnerability(filter.Vulnerability)
```
Comment (lumjjb) [inline pkg/assembler/backends/inmem/certifyVuln.go:232]: do we need to check here that we skip novulns

File: pkg/assembler/backends/inmem/certifyVuln_test.go [hunk 4]
```diff
@@ -355,6 +355,89 @@ func TestIngestCertifyVulnerability(t *testing.T) {
...
+					CertifyVuln: &model.ScanMetadataInput{
+						Collector:      "test collector",
+						Origin:         "test origin",
+						ScannerVersion: "v1.0.0",
+						ScannerURI:     "test scanner uri",
+						DbVersion:      "2023.01.01",
+						DbURI:          "test db uri",
+						TimeScanned:    t1,
+					},
+				},
```
Comment (lumjjb) [inline pkg/assembler/backends/inmem/certifyVuln_test.go:375]: can add another input here as well to test that it doesn't tkae the additional?

File: pkg/assembler/backends/inmem/certifyVuln_test.go [hunk 5]
```diff
@@ -355,6 +355,89 @@ func TestIngestCertifyVulnerability(t *testing.T) {
...
+				{
+					Package: p1out,
+					Vulnerability: &model.Vulnerability{
+						Type:             "cve",
+						VulnerabilityIDs: []*model.VulnerabilityID{c1out},
+					},
+					Metadata: vmd1,
+				},
+			},
+		},
```
Comment (lumjjb) [inline pkg/assembler/backends/inmem/certifyVuln_test.go:440]: can we add a test for novuln not set withn both novuln and 2 vuln inputs?

**Review by lumjjb — APPROVED** [review]
minor nits / testing requests, else LGTM

File: pkg/assembler/backends/inmem/certifyVuln.go (hunk 3 above)
Comment (pxp928) [reply pkg/assembler/backends/inmem/certifyVuln.go:232]: addded check in the resolver such that if `novuln` is set to false, the type cannot be set to "novuln"

Comment (pxp928) [reply pkg/assembler/backends/inmem/certifyVuln.go:232]: added a check here and in vulnerabilities to check also.

File: pkg/assembler/backends/inmem/certifyVuln_test.go (hunk 4 above)
Comment (pxp928) [reply pkg/assembler/backends/inmem/certifyVuln_test.go:375]: added! Only returns the novuln nodes

File: pkg/assembler/backends/inmem/certifyVuln_test.go (hunk 5 above)
Comment (pxp928) [reply pkg/assembler/backends/inmem/certifyVuln_test.go:440]: added test with novuln boolean omitted and it returns all vulnerabilities (including novuln)

File: pkg/assembler/backends/inmem/certifyVuln.go (hunk 3 above)
Comment (pxp928) [reply pkg/assembler/backends/inmem/certifyVuln.go:232]:
```
query CertifyNoVuln {
  CertifyVuln(certifyVulnSpec: {
    vulnerability: {noVuln: false, type: "novuln"}
  }) {
    ...allCertifyVulnTree
  }
}
```
will return in error:
```
{
  "errors": [
    {
      "message": "novuln boolean set to false, cannot specify vulnerability type to be novuln",
      "path": [
        "CertifyVuln"
      ]
    }
  ],
  "data": null
}
```

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**